src/dbami/db.py

```diff
@@ -406,20 +402,85 @@ async def _update_schema_version(
             await self.execute_sql(schema_sql, conn=conn)
             await self.execute_sql(table_sql, conn=conn)
 
+    async def _update_schema_version(
+        self,
+        version: int,
+        conn: asyncpg.Connection,
+    ) -> None:
+        await self._create_schema_and_version_table(conn)
+
+        version_sql, params = render(
+            "INSERT INTO :version_table (version) VALUES (:version)",
+            version_table=V(self.schema_version_table),
+            version=version,
+        )
+
         await self.execute_sql(version_sql, *params, conn=conn)
 
+    @asynccontextmanager
+    async def migration_lock(
+        self,
+        use_lock: bool = True,
+        timeout_ms: int = 0,
+        **kwargs,
+    ) -> AsyncIterator[asyncpg.Connection]:
+        async with self.get_db_connection(**kwargs) as conn:
+            if not use_lock:
+                yield conn
+                return
+
+            await self._create_schema_and_version_table(conn)
+
+            lock_query, _ = render(
+                """
+                DO $_$
+                    BEGIN
+                        SET lock_timeout = :timeout_ms;
+                        PERFORM pg_advisory_lock(
+                            :dbami_lock_id,
+                            to_regclass(':version_table')::integer
+                        );
+                    END
+                $_$;
+                """,
+                timeout_ms=V(str(timeout_ms)),
+                dbami_lock_id=V(str(DBAMI_LOCK_ID)),
+                version_table=V(self.schema_version_table),
+            )
+            unlock_query, unlock_params = render(
+                """
+                SELECT pg_advisory_unlock(
+                    :dbami_lock_id,
+                    to_regclass(:version_table)::oid::integer
+                )
+                """,
+                dbami_lock_id=DBAMI_LOCK_ID,
+                version_table=self.schema_version_table,
+            )
+
+            try:
+                await self.execute_sql(lock_query, conn=conn)
+            except asyncpg.exceptions.LockNotAvailableError:
+                raise exceptions.LockError(
```

Comment:
I don't see this caught anywhere else, so a caller of this must handle this themselves? I think that should be documented somewhere. Though, I also hate exceptions, and would prefer that the entry point to all of this code use a Result pattern to return a Success(value) or Failure(exception).

Reply:
Correct, a caller would have to handle the LockError themselves. Regarding documentation, are you thinking in the method docstring? There are no docs for this repo.

Side note, this prompted me to add the error to those caught by the CLI for the purpose of outputting a the custom exception message (more user friendly) rather than a traceback.